tests/test_surface_events.cpp

```diff
@@ -495,6 +495,49 @@ TEST_F(ClientSurfaceEventsTest, surface_moves_while_under_pointer)
         });
 }
 
+TEST_F(ClientSurfaceEventsTest, frame_timestamp_increases)
+{
+    using namespace testing;
+
+    wlcs::Client client{the_server()};
+
+    auto surface = client.create_visible_surface(100, 100);
+
+    std::array<wlcs::ShmBuffer, 3> buffers = {{
+        wlcs::ShmBuffer{client, 100, 100},
+        wlcs::ShmBuffer{client, 100, 100},
+        wlcs::ShmBuffer{client, 100, 100}
+    }};
+
+    /*
+     * The first buffer must never be released, since it is replaced before
+     * it is committed, therefore it never becomes busy.
+     */
+    wl_surface_attach(surface, buffers[0], 0, 0);
+    wl_surface_attach(surface, buffers[1], 0, 0);
+
+    int prev_frame_time = 0;
+    int frame_callback_count = 0;
+    surface.add_frame_callback(
+        [&](int time)
+        {
+            EXPECT_THAT(time, Gt(prev_frame_time));
+            prev_frame_time = time;
+            frame_callback_count++;
+        });
+    wl_surface_commit(surface);
+
+    usleep(10000);
```

Comment:
Adding sleeps in the middle of tests smells badly of race conditions.

Why? And isn't there a better way?

Reply:
The reason is that we are testing that a millisecond resolution timestamp actually increases as time passes. Therefor we have to wait multiple milliseconds for the test to make sense.

Reply:
The reason is that we are testing that a millisecond resolution timestamp actually increases as time passes. Therefor we have to wait multiple milliseconds for the test to make sense. I will add a comment explaining how it works in more detail.

Reply:
It's still a shame we can't simply provide a test timesource run independently of real time.